### PR TITLE
Port mtom support from dotnet/wcf

### LINKED ISSUE #10: Add support for MTOM message encoding
Want to configure a custom binding SOAP 1.2 with MTOM support. Is this possible

### PR DESCRIPTION
**Pull request by g7ed6e**
Resolves #10 .

### CODE REVIEW

**Comment by birojnayak** [comment]
thank you ! did a quick look, assuming you still need to add some unit tests...

**Comment by g7ed6e** [comment]
> thank you ! did a quick look, assuming you still need to add some unit tests...

Indeed. But I can’t find SSM 4.9.0-rc1 on nuget.org. Do you know if i can get it somewhere else ?

**Comment by mconnew** [comment]
@g7ed6e, I don't know if the release date is public so I can't tell you when 4.9.0-rc1 will be released, but it is very soon.

**Comment by mconnew** [comment]
The packages are now available. So you should be able to write tests now.

File: src/CoreWCF.Http/tests/ContractBehaviorTests.cs [hunk 1]
```diff
@@ -246,7 +246,8 @@ private void Variation_ByHandImplementsOther()
                 cf.Endpoint.Contract.ContractBehaviors.Add(theBehavior);
                 cf.Open();
                 string expected = "IContractBehavior:ClientContract.MyMultiFacetedBehaviorAttribute;";
-                Assert.Equal(expected, BehaviorInvokedVerifier.ValidateClientInvokedBehavior(cf.Endpoint));
+                string actual = BehaviorInvokedVerifier.ValidateClientInvokedBehavior(cf.Endpoint);
+                Assert.Equal(expected, actual);
```
Comment (g7ed6e) [inline src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: this assertion is failing when running the Theory, but does not when running only this test case. I did not get why until now.

Comment (g7ed6e) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: @mconnew The SSM package update to rc1 is causing this issue. You can reproduce by updating the SSM packages in main branch.

Comment (g7ed6e) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: Please see this draft PR that repro the issue https://github.com/CoreWCF/CoreWCF/pull/450

Comment (mconnew) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: Oh dear, I think I might have screwed something up in ServiceContract when I added some caching. I'll investigate and update you.

Comment (mconnew) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: Yep, I did. I added a feature to the client to cache the ContractDescription between instantiations of the type for the same contract. I forgot that behaviors are held by the description, which means it's caching the behaviors too. So when using theories, it creating multiple instances of ChannelFactory, one for each variation, and retaining the behaviors. You should be able to work around it for now by doing `GC.Collect(2)` twice before you instantiate the ChannelFactory. I will fix this before we release.

Comment (g7ed6e) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: Fixed in #450

Comment (mconnew) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: I've merged #450 now.

Comment (g7ed6e) [reply src/CoreWCF.Http/tests/ContractBehaviorTests.cs:250]: Thanks please assign #459 to me.

File: src/CoreWCF.Http/src/CoreWCF/WSMessageEncoding.cs [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
+    internal static class WSMessageEncodingHelper
+    {
+        internal static bool IsDefined(WSMessageEncoding value)
+        {
+            return
+                value == WSMessageEncoding.Text
+                || value == WSMessageEncoding.Mtom;
+        }
+        internal static void SyncUpEncodingBindingElementProperties(TextMessageEncodingBindingElement textEncoding, MtomMessageEncodingBindingElement mtomEncoding)
+        {
```
Comment (mconnew) [inline src/CoreWCF.Http/src/CoreWCF/WSMessageEncoding.cs:23]: Nit, need blank space between methods

File: src/CoreWCF.Http/tests/MessageEncodingTests.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        {
+            IWebHost host = ServiceHelper.CreateWebHostBuilder(_output, startupType).Build();
+            using (host)
+            {
+                host.Start();
+                var factory = new System.ServiceModel.ChannelFactory<ClientContract.IMessageEncodingService>(binding,
+                    new System.ServiceModel.EndpointAddress(new Uri("http://localhost:8080/MessageEncodingService/IMessageEncodingService.svc")));
+                ClientContract.IMessageEncodingService channel = factory.CreateChannel();
+                byte[] bytes = new byte[768];
+#if NET472_OR_GREATER
```
Comment (mconnew) [inline src/CoreWCF.Http/tests/MessageEncodingTests.cs:49]: There's some heuristics around whether to encode a chunk separately or put it inline. I have a vague memory of size being a factor. Can you verify that this does actually place the request and response byte arrays in a separate multipart section? If might be worth adding a test for a large request (eg 1MB) and making this one small, like 10 bytes.

File: src/CoreWCF.Http/tests/MessageEncodingTests.cs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+                {
```
Comment (mconnew) [inline src/CoreWCF.Http/tests/MessageEncodingTests.cs:51]: I can see this being a common thing to use in tests, can you refactor this into the one of the helper classes to a method which returns a random byte array of the specified size.

File: src/CoreWCF.Primitives/src/CoreWCF/Channels/MtomMessageEncodingBindingElement.cs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentException(SR.Format(SR.MtomEncoderBadMessageVersion, value.ToString()), nameof(value)));
+                }
+
+                _messageVersion = value;
+            }
+        }
+
+        // public override IChannelFactory<TChannel> BuildChannelFactory<TChannel>(BindingContext context)
+        // {
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Channels/MtomMessageEncodingBindingElement.cs:155]: Can you remove these two commented methods

**Comment by mconnew** [comment]
>         public override ICollection ReadContentAsList()

This is dead code, please remove

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:898 in 4270416. [](commit_id = 4270416dc23f90d518d4f7688df916622304a124, deletion_comment = False)

**Comment by mconnew** [comment]
>             public override bool IsStartSubsetElement()

Deal code, please remove

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:1593 in 4270416. [](commit_id = 4270416dc23f90d518d4f7688df916622304a124, deletion_comment = False)

**Comment by mconnew** [comment]
>                 get { throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.Format(SR.SeekNotSupportedOnStream, GetType().FullName))); }

I believe these two pragma's can be removed, this one and the one in the next method

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:2157 in 4270416. [](commit_id = 4270416dc23f90d518d4f7688df916622304a124, deletion_comment = False)

**Comment by mconnew** [comment]
>         [SecuritySafeCritical]

Remove the commented out attribute and the other two attributes. These were for static analysis in the original .NET Framework build system, and Code Access Security (CAS), neither of which are applicable now.

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:2689 in 4270416. [](commit_id = 4270416dc23f90d518d4f7688df916622304a124, deletion_comment = False)

**Comment by mconnew** [comment]
>                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.Format(SR.SeekNotSupportedOnStream, stream.GetType().FullName)));

Remove this pragma and the one in the method below.

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:2886 in 4270416. [](commit_id = 4270416dc23f90d518d4f7688df916622304a124, deletion_comment = False)

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs [hunk 6]
```diff
@@ -0,0 +1,1550 @@
...
+            await _mimeWriter.StartPartAsync();
+            if (contentID != null)
+                _mimeWriter.WriteHeader(MimeGlobals.ContentIDHeader, string.Format(CultureInfo.InvariantCulture, "<{0}>", contentID));
+            if (contentTransferEncoding != null)
+                _mimeWriter.WriteHeader(MimeGlobals.ContentTransferEncodingHeader, contentTransferEncoding);
+            if (contentType != null)
+                _mimeWriter.WriteHeader(MimeGlobals.ContentTypeHeader, contentType);
+        }
+#if NO
+        public override bool CanSubsetElements
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:600]: Remove dead code

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs [hunk 7]
```diff
@@ -0,0 +1,1550 @@
...
+                _writer = _infosetWriter;
+                this._infosetWriter = null;
+                _contentTypeStream = null;
+                if (_contentType != null)
+                    Writer.WriteString(_contentType);
+            }
+        }
+
+#if NO
+        public override bool ElementSubsetting
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:670]: Remove dead code

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs [hunk 8]
```diff
@@ -0,0 +1,1550 @@
...
+            Writer.WriteComment(text);
+        }
+
+        public override void WriteDocType(string name, string pubid, string sysid, string subset)
+        {
+            WriteBase64InlineIfPresent();
+            Writer.WriteDocType(name, pubid, sysid, subset);
+        }
+#if NO
+        public override void WriteElementSubset(ArraySegment<byte> buffer)
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:754]: Dead code

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs [hunk 9]
```diff
@@ -0,0 +1,1550 @@
...
+        }
+
+        public override void WriteValue(long value)
+        {
+            WriteBase64InlineIfPresent();
+            Writer.WriteValue(value);
+        }
+
+#if DECIMAL_FLOAT_API
+        public override void WriteValue(decimal value)
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:999]: These two methods are deal code, please remove

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs [hunk 10]
```diff
@@ -0,0 +1,1550 @@
...
+                char c = value[index + i];
+                if ((ushort)c > 0xFF)
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new FormatException(SR.Format(SR.MimeHeaderInvalidCharacter, c, ((int)c).ToString("X", CultureInfo.InvariantCulture))));
+                buffer[offset + i] = (byte)c;
+            }
+            offset += count;
+        }
+
+#if NO
+        internal void Write(byte value)
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:1476]: This is dead code, please remove

**Comment by mconnew** [comment]
It's looking good. There's some code cleanup needed which I know was in the wcf client repo, and I'd like the tests to test a small and large request, but other than that, this looks good. It's very close to being mergeable.

File: src/CoreWCF.Http/tests/MessageEncodingTests.cs (hunk 3 above)
Comment (g7ed6e) [reply src/CoreWCF.Http/tests/MessageEncodingTests.cs:49]: Indeed. Threshold is 767.
```
// Maximum number of bytes that are inlined as base64 data without being MTOM-optimized as xop:Include
private const int MaxInlinedBytes = 767;  // 768 will be the first MIMEd length
```

File: src/CoreWCF.Http/tests/MessageEncodingTests.cs [hunk 11]
```diff
@@ -189,5 +203,51 @@ protected override Channels.Binding GetServerBinding()
                 return binding;
             }
         }
+
+        internal class AssertMtomOptimizedEndpointBehvaior : System.ServiceModel.Description.IEndpointBehavior
+        {
+            class AssertMtomOptimizedDelegatingHandler : DelegatingHandler
```
Comment (g7ed6e) [inline src/CoreWCF.Http/tests/MessageEncodingTests.cs:208]: I first tried to use a `IClientMessageInspector.AfterReceiveReply` to inspect the response client side but the provided Message was already 'rebuilt' hiding the multipart.

File: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs (hunk 8 above)
Comment (mconnew) [reply src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomWriter.cs:736]: You missed this one in the dead code cleanup.

**Comment by mconnew** [comment]
>             }

You can actually go one step further and do this:
```c#
public override string LocalName => _readState == ReadState.Interactive ? string.Empty : _parentReader.LocalName;
```

---
Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:1590 in 3ff896e. [](commit_id = 3ff896e678bd6bccd229bc01d344f129a567b9b2, deletion_comment = False)

**Comment by mconnew** [comment]
There's the one bit of dead code you missed. After that's fixed, this is ready to merge.

**Comment by g7ed6e** [comment]
> > ```
> >         }
> > ```
> 
> You can actually go one step further and do this:
> 
> ```cs
> public override string LocalName => _readState == ReadState.Interactive ? string.Empty : _parentReader.LocalName;
> ```
> 
> Refers to: src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlMtomReader.cs:1590 in [3ff896e](https://github.com/CoreWCF/CoreWCF/commit/3ff896e678bd6bccd229bc01d344f129a567b9b2). [](commit_id = [3ff896e](https://github.com/CoreWCF/CoreWCF/commit/3ff896e678bd6bccd229bc01d344f129a567b9b2), deletion_comment = False)

I cleaned both `XmlMtomReader` and `XmlMtomWriter`. please see https://github.com/CoreWCF/CoreWCF/pull/442/commits/e27fcbde8e3557198f1f43a571aa58edeace89f0

**Comment by g7ed6e** [comment]
> There's the one bit of dead code you missed. After that's fixed, this is ready to merge.

Thanks for seeing it. I double checked `#if NO` and found other dead code. please see https://github.com/CoreWCF/CoreWCF/pull/442/commits/02103bc69fa8bbc4652ade5d8c396c6766346634

**Comment by g7ed6e** [comment]
@mconnew this one should be good now.

**Review by mconnew — APPROVED** [review]

**Pull request merged**